llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp

```diff
@@ -1228,6 +1228,10 @@ void RISCVDAGToDAGISel::Select(SDNode *Node) {
       if (!Subtarget->hasVendorXTHeadBb())
         return false;
 
+      if ((Subtarget->hasStdExtZba() || Subtarget->hasStdExtZbb()) &&
+          (Msb == 7 || Msb == 15 || Msb == 31) && Lsb == 0)
+        return false;
```

Comment:
This, of course, assumes that Zba/Zbb instructions have the same or lower cost than XTheadBB instructions, if implemented on the same core.